src/types.md

```diff
@@ -236,3 +236,311 @@ fn main() {
 
 Of course, enumerating all possible stored variants remains preferable such that the
 compiler helps you to avoid runtime panics.
+
+## When should I put my data in a `Box`?
+
+In C++, you often need to box things for ownership reasons, whereas in Rust
+it's typically just a performance trade-off. It's arguably premature optimization
+to use boxes unless your profiling shows a lot of memcpy of that particular
+type (or, perhaps, the relevant [clippy lint](https://rust-lang.github.io/rust-clippy/v0.0.212/index.html#large_enum_variant)
+informs you that you have a problem.)
+
+> I never box things unless it’s really big - MG
+
+Another heuristic is if part of your data structure is very rarely filled,
+in which case you may wish to `Box` it to avoid incurring an overhead for all
+other instances of the type.
+
+```rust
+# struct Humility; struct Talent; struct Ego;
+struct Popstar {
+  ego: Ego,
+  talent: Talent,
+  humility: Option<Box<Humility>>,
+}
+# fn main() {}
+```
+
+(This is one reason why people like using [anyhow](https://docs.rs/anyhow/latest/anyhow/)
+for their errors; it means the failure case in their `Result` enum is only
+a pointer wide.)
+
+Of course, Rust may require you to use a box:
+
+* if you need to `Pin` some data, typically for async Rust
+* if you otherwise have an infinitely sized data structure.
+
+but as usual, the compiler will explain very nicely.
+
+## I miss operator overloading! What do I do?
+
+Implement the standard traits instead. This has equivalent effect in that
+folks will be able to use your type in a standard Rusty way without knowing
+too much special about your type.
+
+## Should I have public fields or accessor methods?
+
+The trade-offs are similar to C++ except that Rust's pattern-matching makes it
+very convenient to match on fields, so within a realm of code that you own you
+may bias towards having more public fields than you're used to. As with C++,
+this can give you a future compatibility burden.
+
+## When should I use a newtype wrapper?
+
+The [newtype wrapper pattern](https://rust-unofficial.github.io/patterns/patterns/behavioural/newtype.html)
+uses Rust's type systems to enforce extra behavior without necessarily changing
+the underlying representation.
+
+```rust
+# fn get_rocket_length() -> Inches { Inches(7) }
+struct Inches(u32);
+struct Centimeters(u32);
+
+fn build_mars_orbiter() {
+  let rocket_length: Inches = get_rocket_length();
+  // mate_to_orbiter(rocket_length); // does not compile because this takes cm
+}
+```
+
+Other examples that have been used:
+* An IP address which is guaranteed not to be localhost;
+* Non-zero numbers;
+* IDs which are guaranteed to be unique
+
+Such new types typically need a lot of boilerplate, especially to implement
+the traits which users of your type would expect to find. On the other hand,
+they allow you to use Rust's type system to statically prevent logic bugs.
```

Comment:
Maybe cross link this answer with the "when should I use static checking" question. I like internal cross-links!